mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -274,6 +274,82 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
   }
 };
 
+template <typename ArithOp, bool needsUnsigned>
+class CastConversion : public OpConversionPattern<ArithOp> {
+public:
+  using OpConversionPattern<ArithOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArithOp op, typename ArithOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type opReturnType = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType, IndexType>(opReturnType)) {
+      return rewriter.notifyMatchFailure(op, "expected result integer type");
+    }
+
+    if (adaptor.getOperands().size() > 1) {
+      return rewriter.notifyMatchFailure(
+          op, "CastConversion only supports unary ops");
+    }
+
+    Type operandType =
+        this->getTypeConverter()->convertType(adaptor.getIn().getType());
+    if (!isa_and_nonnull<IntegerType, IndexType>(operandType)) {
+      return rewriter.notifyMatchFailure(op, "expected operand integer type");
+    }
+
+    bool doUnsigned =
+        needsUnsigned ||
+        (isa<IntegerType>(operandType) && isa<IntegerType>(opReturnType) &&
+         operandType.getIntOrFloatBitWidth() >
+             opReturnType.getIntOrFloatBitWidth());
+
+    Type castType = opReturnType;
+    // For int conversions: if the op is a ui variant and the type wanted as
+    // return type isn't unsigned, we need to issue an unsigned type to do
+    // the conversion.
+    if (isa<IntegerType>(opReturnType) &&
+        castType.isUnsignedInteger() != doUnsigned)
+      castType = rewriter.getIntegerType(opReturnType.getIntOrFloatBitWidth(),
```

Comment:
If I'm doing a truncation, why do I need to convert the result into unsigned? I'm assuming that is what is happening here because in that case, `castType.isUnsignedInteger()` should be false and `doUnsigned` true

Reply:
I'm not sure I understand exactly where you go, so let's see if I answer correctly. This is not to convert the result, this is the actual conversion taking place with this `castType`.

The behavior of C casts depends on both the origin and destination types. If we want to ensure a specific behavior in C (sign-extension, truncation), we need to fix them (hence the operand conversion above and the castType).

Then, we go back to signless because that's what's expected from arith operations.